### PR TITLE
Support STRUCT in UDAFs (both parameters & return values)

### PR DESCRIPTION
**Pull request by agavra**
### Description 
#2642 and #2677 covered Struct support for UDFs. This follows it up by supporting a similar syntax for UDAFs. Namely, `@UdafFactory` may now optionally specify `paramSchema` and `returnSchema` if it accepts/returns a STRUCT, for example:
```java
@UdafFactory(
      description = "returns a struct with (SUM(S->A), SUM(S->B)) over the window",
      paramSchema = "STRUCT<A INTEGER, B INTEGER>",
      returnSchema = "STRUCT<A INTEGER, B INTEGER>")
```

### Testing done 
- Unit testing
- new QTT coverage

### Postscript

Yes... I am in London. I decided to visit the London office and was surprisingly productive in just two or three hours so I figured I'd publish this PR.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/udf.rst [hunk 1]
```diff
@@ -448,17 +480,42 @@ UdafFactory Annotation
...
++-------------+------------------------------+------------------------+
+| returnSchema| The KSQL schema for the      | For complex types      |
+|             | return value.                | such as STRUCT         |
++-------------+------------------------------+------------------------+
+
+.. note:: If ``paramSchema`` or ``returnSchema`` is supplied in the ``@UdfParameter`` annotation for
+          a ``STRUCT`` it is considered "strict" - any inputs must match exactly, including order
+          and names of the fields.
+
+You can use this to better describe what a particular version of the UDF does, for example:
```
Comment (JimGalasyn) [inline docs/developer-guide/udf.rst]: UDAF?

File: docs/changelog.rst [hunk 2]
```diff
@@ -26,6 +26,23 @@ KSQL 5.3.0 includes new features, including:
 * ``CREATE STREAM`` and ``CREATE TABLE`` will now allow you to create the topic if it is missing.
   To do this, specify the ``PARTITIONS`` and optionally ``REPLICAS`` in the ``WITH`` clause.
 
+* KSQL now supports deserializing records where the value is:
```
Comment (vcrfxia) [inline docs/changelog.rst]: The first two bullets don't seem related to this PR. Perhaps they were accidentally picked up in a bad merge?

Comment (agavra) [reply docs/changelog.rst]: yes. seems like a bad merge indeed

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java [hunk 3]
```diff
@@ -136,11 +147,15 @@ public static UdfInvoker compile(final Method method, final ClassLoader loader)
...
-          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.left));
+      final Schema argSchema = paramSchema.isEmpty()
+          ? SchemaUtil.getSchemaFromType(valueAndAggregateTypes.left)
+          : LogicalSchemas.fromSqlTypeConverter().fromSqlType(TypeContextUtil.getType(paramSchema));
+      final List<Schema> args = Collections.singletonList(argSchema);
 
-      final Schema returnValue = SchemaUtil.ensureOptional(
-          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.right));
+      final Schema returnValue = returnSchema.isEmpty()
+          ? SchemaUtil.ensureOptional(SchemaUtil.getSchemaFromType(valueAndAggregateTypes.right))
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java]: Not a change from this PR but I'm curious: how come we ensure optionality of the return schema but not the input schema?

File: ksql-engine/src/test/java/io/confluent/ksql/function/UdfCompilerTest.java [hunk 4]
```diff
@@ -271,98 +275,119 @@ public void shouldThrowWhenUdafReturnTypeIsntAUdaf() throws Exception {
...
-    );
+        classLoader,
+        "test",
+        "desc",
+        "",
+        "");
+  }
+
+  @Test
+  public void shouldHandleUdafsWithStructStructTypes() throws Exception {
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-engine/src/test/java/io/confluent/ksql/function/UdfCompilerTest.java]: Struct for input and Struct for output

File: ksql-udf/src/main/java/io/confluent/ksql/function/udaf/UdafFactory.java [hunk 5]
```diff
@@ -80,5 +80,27 @@
...
  */
 public @interface UdafFactory {
+
+  /**
+   * @return a description for the UDAF
+   */
   String description();
+
+  /**
+   * The schema for the parameter.
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-udf/src/main/java/io/confluent/ksql/function/udaf/UdafFactory.java]: Correct - this is a limitation of the new framework

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java (hunk 3 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java]: if the input schema is required, then a `null` value will be treated as an error (e.g. a `null` value will be returned). On the other hand, all schemas must be optional - which is what the output determines.

File: docs/developer-guide/udf.rst [hunk 6]
```diff
@@ -421,6 +421,38 @@ used to call the UDAF. The UDAF can be invoked in four ways:
         };
       }
 
+      @UdafFactory(
```
[the remark this answers is not in the document]
Comment (agavra) [reply docs/developer-guide/udf.rst]: good catch. updated
